### DISCUSSION TITLE
Towards LobeChat 1.0 | 迈向 LobeChat 1.0

### DISCUSSION THREAD

**Discussion by arvinxx** [lobehub/lobehub#1768] [Announcements]
> English | [简体中文](https://github.com/lobehub/lobe-chat/discussions/1768#discussioncomment-8930495)

## LobeChat 0.x Review

In the version 0.x of LobeChat, adhering to the principle of **"If not necessary, do not add entities"**, we adopted a serverless, Local First implementation plan. This decision was based on three core reasons:

- **Fast Response**: Handling all operations directly from the client can significantly improve response speed.
- **Data Privacy**: The absence of centralized data storage reduces the risk of data leakage or misuse.
- **Simple Deployment**: Eliminating the database means users can start using it more quickly and easily.

It turned out that this approach was very friendly for most scenarios. Our users no longer had to worry about losing data due to being blocked by OpenAI and were able to protect their data privacy well.

However, in practice, we gradually realized that this design limited the upper limit of our product capabilities. In the absence of server-side persistent storage, many features that users urgently needed were difficult or completely impossible to implement. For example:

- **Cross-device message synchronization**: Once users have multiple devices in hand, data synchronization becomes a necessity, which is a highly requested feature within the community. Under the current mode, only relatively unconventional methods (WebRTC) can be used, while having a database would make everything much simpler and more user-friendly.
- **Knowledge base/PDF/document session**: Knowledge base, PDF, or document session functionality is the highest demand from users, and currently, mature solutions in the community mostly rely on cloud services or cloud databases. If implemented purely on the client side, it would be quite difficult and unable to reuse mature community solutions.
- **High-quality experience sharing**: We once tried using a URL to save data, which was barely manageable with a small volume of conversations, but long dialogues far exceeded the information capacity that a URL could hold. Without database support, this greatly limited the circulation of valuable knowledge.
- **Running scheduled tasks/support for assistants to send messages proactively**: If there are long-running tasks, the current pure client-side mode is completely unable to implement this function. We inevitably need assistants that can run silently on the server side to achieve proactive interaction with us.

Therefore, it's time to move forward~ We will introduce **server-side database support** in LobeChat 1.0 to better meet everyone's needs for new features!

## Toward 1.0: Keeping the Open Source Spirit Unchanged

We have decided to officially introduce server-side database support in LobeChat 1.0, laying the infrastructure groundwork for the next phase of LobeChat. However, we had a heated discussion about whether to open-source this part:

On one hand, we want to bring a better product experience to all LobeChat users and further improve efficiency in daily work, making the database an indispensable factor. On the other hand, we pondered whether our commercial version, LobeChat Cloud, would lose its unique competitive edge if everything was open-sourced.

This issue has been unresolved since we started considering commercialization. Recently, reviewing my Cubox, I revisited a phrase that didn't strike me much at the time, but now, standing at this point in time, it inspired me greatly: 

![image](https://github.com/lobehub/lobe-chat/assets/28616219/03ed22f1-09c8-4937-9ae1-f47130da23b0)

The starting point of LobeChat was actually very simple — the AI conversation products on the market were [not good enough](https://twitter.com/arvin17x/status/1768590012583702773), so I wanted to create a version that was comfortable enough for myself to use.

Perhaps a 60-point Chat product might be sufficient for most businesses and individuals, but as designers, we are not satisfied with a 60-point product. We want to take the "AI conversation" scenario to 80 points, 90 points, or even 100 points. Thus, LobeChat was born. Now, we are on the path to "a better AI conversation product," exploring possible interaction modes for the next generation of AI conversation products and making them accessible to everyone.

I believe LobeChat will eventually become a unique and excellently experienced AI conversation product in the world. As for commercial returns, I think it will come naturally.

Therefore, after careful consideration, LobeChat 1.0 will still keep all core functionalities open-source. We are very excited about the Roadmap for LobeChat 1.0 this year and believe LobeChat will become even more amazing this year. We will soon provide a detailed explanation of this year's Roadmap, so you can look forward to it a little.

![image](https://github.com/lobehub/lobe-chat/assets/28616219/aa93ef54-f084-4194-a221-7036b58ad9b7)

## 1.0 License Adjustment for Healthier Development
The development of LobeChat would not be possible without the encouragement and support of the community. Due to the deep involvement of the community, our product has been able to grow and improve continuously. Since the project's inception, we have been using the MIT license, which has greatly promoted the project's early development and technology sharing. As our product evolves and our community grows, we started thinking about how to better protect the interests of the project, contributors, and users while promoting the healthy development of our product ecosystem.

After thorough consideration and preliminary discussions with community members, we have decided to update the open-source license of LobeChat from the MIT license to the Apache 2.0 license starting from 1.0, and add a commercialization authorization supplementary clause. We believe this change will have the following positive impacts on our project:

1. **Stronger Patent Protection**: The Apache License 2.0 provides clearer patent licenses for contributors and users, meaning anyone using or distributing our product will automatically receive a license for the relevant patents. This is important for protecting our community's innovative achievements and preventing patent litigation risks.
2. **Clear Contributor Responsibilities**: Through the Apache License, we can more clearly define the responsibilities and obligations of contributors, ensuring that every community member's contribution is made in a safe and respectful environment.
3. **Encouraging Commercial Use**: The Apache License encourages commercial entities to use and contribute to open-source projects without worrying about excessive legal risks. This is positive for attracting more corporate users to participate in our project, expanding our product's application scenarios and user base.

For most users, this license change will have no impact. For the following scenarios, LobeChat will remain completely free:

- All personal users / teams using it internally, self-deployed scenarios, remain free to use;
- Second development on LobeChat without commercialization (e.g., internal company use, providing public welfare services), remain free;
- Direct use of the original version of LobeChat for multi-tenant commercialization without any modifications remains free (e.g., various API relay stations, using LobeChat as the client UI);

Only when conducting secondary development on LobeChat and commercializing it (e.g., hiding the official logo, modifying official links) will you need to obtain commercialization authorization from us. If you have such needs, you can contact us by email: hello@lobehub.com.

We are well aware that only with a prosperous community can LobeChat's vitality continue. Therefore, we will try to open a "free commercialization authorization plan" for open-source contributors in the future. As long as you submit a PR to LobeChat, we will directly grant commercialization authorization for free based on the quality of the PR, encouraging commercial entities to actively participate in the construction of LobeChat's community ecosystem.

Toward LobeChat 1.0, let's build a community-based open-source product together and explore the boundaries of AI's future!

**Comment by arvinxx** [comment]
> [English](https://github.com/lobehub/lobe-chat/discussions/1768#discussion-6428235) | 简体中文


## LobeChat 0.x 回顾

在LobeChat 0.x 版本中，秉承着 **「如无必要，勿添实体」** 的原则，我们采取了一个无服务端数据库的、 Local First 的实现方案。这个决定基于三个核心原因：

- **响应迅速**：直接从客户端处理所有操作可显著提高响应速度。
- **数据隐私**：没有集中存储数据减少了数据泄露或被滥用的风险。
- **部署简单**：省去数据库意味着用户可以更快速、轻松地开始使用。

事实证明这个思路对于大部分场景来说都非常友好，我们的用户再也不用担心被 OpenAI 封号导致数据丢失的问题，也很好的保护了自己的数据隐私。

然而，在实践过程中，我们逐渐认识到这种设计限制了我们的产品能力上限。在缺乏服务端持久化存储的情况下，许多用户迫切需要的功能实现困难，或完全无法实现。比如：

- **跨端消息同步 ：** 一旦用户手上存在多台设备，就势必需要数据同步，也是社区中用户呼声非常高的部分。而在当前的模式下，只能采用相对不常规的方式来实现（WebRTC），而如果有了数据库，一切就都会变得更加简单与易用。
- **知识库/PDF/文档会话：** 知识库、PDF或者文档会话功能是用户呼声最高的需求了，而目前社区成熟的方案基本都依赖了云端服务或者云端数据库。如果用纯客户端的方式实现难度比较大，也没法复用社区成熟的方案。
- **高质量体验分享：** 之前曾尝试过使用 url 保存数据的方案，在较少会话量的情况下还可堪一用，但超长对话就远远超过了 URL 能承载的信息量，在缺乏数据库支持的情况下，大大局限了有价值的知识流通。
- **运行定时任务/助手主动发送消息支持**： 如果存在需要长时运行的任务，当前纯客户端的模式就完全无法实现该功能。我们势必需要可以在服务端默默运行的助手，这样的助手才能主动与我们交互。

因此，是时候向前迈进了~ 我们将会在 LobeChat 1.0 中引入**服务端数据库支持**，以更好地满足大家对新功能的需求！

## Toward 1.0： 保持开源初心不变

我们决定在 LobeChat 1.0 中正式引入服务端数据库的支持，进而为下一阶段的 LobeChat 做好基础设施层面的铺垫。不过此时，我们却对是否将这部分开源产生了激烈的讨论：

一方面，我们希望能为所有喜爱 LobeChat 的用户带来更好的产品体验，并更进一步提升日常工作中的效率，数据库是不可或缺的因素。而另一方面，我们又在思考如果全部开源，我们的商业化版本 LobeChat Cloud 独特的竞争力是否会因此丧失。

这个问题从我们开始有商业化的想法之后，就一直没有一个定论。最近在我的 Cubox 中回顾到以前看到的一句话，当时触动不大，而站在当下的时间节点来看，却有莫大的启发：

![image](https://github.com/lobehub/lobe-chat/assets/28616219/03ed22f1-09c8-4937-9ae1-f47130da23b0)

LobeChat 的出发源点其实很简单 —— 市面上的 AI 会话产品[不够好](https://twitter.com/arvin17x/status/1768590012583702773)，所以我就想着做一个让自己用着足够舒服的版本。

可能对于大部分企业和个人来说 60 分的 Chat 产品可能就够用了，但作为设计师出身，我们对 60 分的产品并不满足，我们想把「AI 对话」这个场景，做到 80 分、 90 分乃至 100 分。然后， LobeChat 就诞生了。而现在，我们也走在「更好的 AI 对话的产品」这条路上，探索属于下一代 AI 会话产品可能的交互模式，并为所有人可用。我相信 LobeChat 终将会变成世界上一个独树一帜、体验优良的 AI 会话产品。

因此，经过我们的慎重考虑，**LobeChat 1.0 的所有核心的功能仍然会保持开源！**

对于今年 LobeChat 1.0 的 Roadmap，我们都感到非常兴奋，相信 LobeChat 在今年会变得更加惊艳。我们也会很快给出今年 Roadmap 的详解，大家可以小小的期待一下。

![image](https://github.com/lobehub/lobe-chat/assets/28616219/aa93ef54-f084-4194-a221-7036b58ad9b7)


## 1.0 License 调整，为更健康的发展

LobeChat 的发展离不开社区所有人的鼓励与支持，因为社区的深度参与，我们的产品才能够不断成长和完善。自项目启动以来，我们一直采用 MIT 许可证，这一选择大大促进了项目的早期发展和技术共享。随着我们产品的发展迭代和社区的壮大，我们开始思考如何更好地保障项目、贡献者和用户的利益，同时促进我们产品生态的健康发展。

经过深思熟虑和与社区成员的初步讨论，我们决定从 1.0 开始，将 LobeChat 的开源协议从 MIT 许可证更新为 Apache 2.0 协议，并增加一项商业化授权补充条款。我们相信，这一变更将为我们的项目带来以下几方面的积极影响：

1.  **更强的专利保护**：Apache License 2.0 为贡献者和用户提供了更为明确的专利授权，这意味着任何人使用、分发我们的产品，都将自动获得相关专利的授权。这一点对于保护我们社区的创新成果，防止专利诉讼风险具有重要意义。 
2.  **明确的贡献者责任**：通过 Apache License，我们能够更清晰地界定贡献者的责任和义务，确保每一位社区成员的贡献都在一个安全和尊重的环境下进行。 
3.  **促进商业使用**：Apache License 鼓励商业实体使用和贡献开源项目，而不必担心过度的法律风险。这对于吸引更多的企业用户参与我们的项目，扩大我们产品的应用场景和用户基础，具有积极的推动作用。 

对于大部分用户来说，本次协议变更没有任何影响，对于以下场景，LobeChat 仍然保持完全免费：

- 所有个人用户 / 团队内部使用，自行部署的场景，保持免费使用；
- 在 LobeChat 上进行二次开发但不用于商业化的场景（例如企业内部使用、提供公益服务等），保持免费；
- 直接使用原版 LobeChat 进行多租户商业化，不做任何修改的话，保持免费 (例如各类 API 中转站，使用 LobeChat 作为客户端 UI ) ；

只有当在 LobeChat 上进行二次开发并进行商业化（例如隐去官方 Logo 、修改官方链接等），需要向我们获取商业化授权。如有你存在相应的诉求，可以邮件联系：hello@lobehub.com 。

我们深知只有社区足够繁荣，LobeChat 的生命力才会持续下去。因此我们将会在后续尝试开放一项针对开源贡献者的「免费商业化授权计划」，只要向 LobeChat 提交 PR ，我们会根据 PR 质量直接免费发放商业化授权，以此鼓励商业实体积极参与 LobeChat 的社区生态的建设。

Toward LobeChat 1.0，让我们共同构建社区型开源产品，一起探索 AI 的未来边界！

**Reply by loyaniu** [reply]
很好用！特别感谢！辛苦～

**Comment by linth2005** [comment]
可惜我只会纯后台，否则也可以PR一下

**Reply by arvinxx** [reply]
要是会TS版的后端，可以来共建呀~ 😆 后续我们计划是引入 prisma 和 postgres 作为 LobeChat 的数据库

**Reply by mqhe2007** [reply]
盲猜一下是Nest？Nest+prisma+PgSQL是个很好的选择。

**Reply by arvinxx** [reply]
可能不一定用 nest，如果调研下来纯 serverless + prisma 就够的话

**Comment by LeoQuote** [comment]
Looking forward to the lobe chat with simple and quick share function! The code quality is much better than other products, 93% coverage, beautiful graphics, really appreciated all the developers’ work!

**Comment by Sen-CaPoo** [comment]
隨著 LobeChat 即將迎來 1.0 版本的重大里程碑，我藉此機會表達我最深切的感謝。自從我們小團隊開始使用 LobeChat 作為我們的 LLM 主要溝通服務以來，我們發現它不僅提高了 LLM 使用效率，也加深了團隊間對於 LLM 的認識。

LobeChat 的開源精神，讓我們得以見證一個產品從雛形到成熟的過程，這個過程中充滿了來自全球各地的智慧與熱情。每一位貢獻者的辛勞與創意，讓 LobeChat 得以進化成為今日這樣強大而靈活的工具。

隨著 1.0 版本的預告，我們滿懷期待地看著 LobeChat 未來的發展。我相信，隨著社群的持續成長與貢獻，LobeChat 將會變得更加強大，更能滿足我們如今甚至未來的需求。

再次，我要向所有投入時間、精力於 LobeChat 發展的每一位貢獻者表示最深的謝意。正是因為有了你們的貢獻，我們才能擁有這麼出色的 LLM 工具。我們期待與 LobeChat 繼續成長。

**Comment by Root-FTW** [comment]
I could write a huge text thanking you and telling a boring story of how I got to lobechat and I don't plan to leave ever, but I will just say

THANKS FOR EVERYTHING

**Comment by tyff9527** [comment]
非常感谢参与开发的这些人，LobeChat自从部署了之后天天都会用，真的很好！

**Comment by A1den23** [comment]
感谢参与开发的人，一直天天用来改论文

**Comment by mushan0x0** [comment]
希望 LobeHub 大卖 ❤️

**Comment by Zach677** [comment]
感谢每一个贡献参与的人！！

**Comment by cklogic** [comment]
”开源的生命力并不在于“开源”，而在于为人类好好的、真正的解决掉一个问题。“

说得真好！

**Comment by HayGry** [comment]
我想知道 1.0什么时候到来 加油

**Comment by sunsky89757** [comment]
设计师会写代码，简直就是天选之子哟~

**Comment by hjw21century** [comment]
界面简介美观，支持未来更多特性。

**Comment by PearFarmer** [comment]
Thanks so much for the team's decisions on this project! Let's continue to build the best AI tool on Earth!
